### DISCUSSION TITLE
Integrated Graphics Upscaling/Super-resolution

### DISCUSSION THREAD

**Discussion by SequoiaDu** [OpenRCT2/OpenRCT2#15916] [Feature requests]
Since it is almost impossible to recreate all the graphics used in RCT and RCT2, it might be a good idea to use existing upscaling algorithms, such as [ACNet](https://github.com/TianZerL/ACNetGLSL) or [FSR](https://github.com/GPUOpen-Effects/FidelityFX-FSR), in order to smooth the pixelated graphics. There is a project [Magpie](https://github.com/Blinue/Magpie) that works well with several implemented algorithms. It captures a window content and upscales it to fullscreen using a selected method. 

I tried to use Magpie with windowed OpenRCT2 and it worked ok, but with two problems: (1) Menus and texts become too big. (2) It does not work well with different zoom levels. I found there's one zoom level that makes the graphics _quite_ impressive, but others resulted in either too blurry or pixelated graphics.

Integrating an implementation into OpenRCT2 to upscale/smooth the rendered graphics will be great.

**Comment by pizza2004** [comment]
It is my understanding that the need to replace the graphics is to allow the program to be run without any need to own the original games.  Since we don't own a license to use the original graphics for free, and the man who owns them has never come forward to offer them as such, I don't believe it would be possible to achieve the goals simply by upscaling graphics, as we could not distribute them.

**Reply by SequoiaDu** [reply]
I thought if the upscaling is done in real-time on the rendered graphics before output on the screen, there would be no need to own the original graphics. --that being said, another route is to make a program that can parse and upscale all the original artworks, which users can run on their own computers based on the files they have already purchased.

**Reply by ocalhoun6** [reply]
> the need to replace the graphics is to allow the program to be run without any need to own the original games

I still have my doubts about whether that would ever be legal. Even if you've replaced *every* asset in the game ... it's still a copy of the game.

For comparison, suppose I used 2D animation to create my own version of _Avengers: Endgame_. Replaced every frame of the video with my own animation, replaced every second of the audio with my own voice acting and my own music and sound effects. Maybe even re-wrote the script from scratch... *But* as long as it's the same characters in the same story, I'm going to get sued by Disney. As long as it's still recognizable as a version of _Avengers: Endgame_, it's still a copyright (and trademark) infringement.

Likewise, suppose I made my own project called OpenPhotoshop, where I take Adobe Photoshop's code, reverse engineer it, refactor it, build it into my own open-source program that does all the same things the same way ... I'd still get sued by Adobe, even if I didn't keep a single line of code verbatim from the original. Just because it's an _approximate_ copy doesn't mean it's not a copy.

Or I could start distributing my own handwritten copies of the Harry Potter books. But the laborious process of writing each copy out by hand in my own handwriting doesn't make it any less of a copyright violation if I were to start distributing them.

So with RCT, as long as OpenRCT2 is recognizably a version of RCT2, it would still be a copyright violation for us to distribute it. We could replace every sprite in the game ... but as long as we're replacing them with our own versions of sprites that look pretty much the same as the original, you're still violating the copyright. Just because our copies are inexact doesn't mean they're not copies. And, of course, there's still _trademark_ issues. The 'Roller Coaster Tycoon' name is trademarked, so as long as we're using the same name, that's a problem.

I'm fairly sure the only way to distribute OpenRCT2 without requiring RCT2 from the user is to change it _so_ much that it's no longer RCT anymore.

>I don't believe it would be possible to achieve the goals simply by upscaling graphics, as we could not distribute them.

Pretty simple solution for this: distribute the upscaled graphics, but don't allow the user to actually use them unless the original files are detected. As long as the original RCT files are still required, it's not fundamentally any different than what we're doing now.

**Reply by SequoiaDu** [reply]
> Pretty simple solution for this: distribute the upscaled graphics, but don't allow the user to actually use them unless the original files are detected. As long as the original RCT files are still required, it's not fundamentally any different than what we're doing now.

Or maybe build the upscaling program into the installer and let it process the user-owned graphics during the installation as an option if the user chooses to do it.

**Reply by pizza2004** [reply]
Sorry, I misunderstood and thought this was a discussion on the repository to replace all of the graphics rather than the main one, and thought you were suggesting distributing graphics that were upscaled rather than talking about real time upscaling.

If they replace all the files from the original game it's perfectly legal to distribute it for free, and that is the end goal.  OpenTTD does it and it has never gotten in any legal trouble.

Upscaling wouldn't be an issue per se but it wouldn't look as good as properly made graphics of a higher resolution, no matter how you do it.  It's just a question of processing power.

> Pretty simple solution for this: distribute the upscaled graphics, but don't allow the user to actually use them unless the original files are detected. As long as the original RCT files are still required, it's not fundamentally any different than what we're doing now.

This is definitely not legal though, and someone could just compile their own executable by tweaking the open source code to ignore the requirement that you have the original files.

**Reply by duncanspumpkin** [reply]
This is not related to the issue. The issue is just about applying a upscaling filters.

**Reply by ocalhoun6** [reply]
> someone could just compile their own executable by tweaking the open source code to ignore the requirement that you have the original files.

Well, then that's _their_ problem and _they_ can get sued for it.

>OpenTTD does it and it has never gotten in any legal trouble.

Yeah ... but it's not like that's legally binding precedent. If you get sued for copyright/trademark infringement, saying 'but OpenTTD' isn't going to get you very far. It may just be that in OpenTTD's case, the copyright owners haven't noticed or chose not to enforce their copyright because they don't think it's worthwhile. OpenRCT2 might not be so lucky.

I'd really urge consultation with an IP lawyer before putting the effort into distributing a stand-alone game.


----

But yeah, I guess we are getting pretty side-tracked from the actual issue at hand.

In that regards, I'd definitely prefer the graphics upscaling to be done as a one-time thing -- not on the fly during gameplay.

Ideally, upscaled sprites could be distributed along with the game. That has a few advantages:
- Don't have to worry about the potential for bugs being introduced by different hardware and different operating systems doing the upscaling slightly differently.
- Don't have any performance hit while playing the game (vs real-time)
- Don't have extended install times (especially on older computers) while the upscaling processes (vs doing it during install)
- Would still be usable by people with older, less-capable hardware (especially vs real-time upscaling)
(The one disadvantage -- besides any potential legal issue -- is that it would increase the size of the install file, leading to longer download times and slightly more hard drive space taken.)

Failing that, one-time upscaling during installation would be better than doing it realtime. Again, because you wouldn't take any performance hit during gameplay and because that would enable much less capable computers to use the upscaled graphics.

I know, personally, I do _not_ have a very good GPU at all, so realtime GPU upscaling would be a no-go for me. But if it's just a one-time thing, even if it has to run for hours, that's fine. I could just let it run overnight and come back to nice upscaled graphics the next day.

**Comment by IntelOrca** [comment]
This has been attempted before:
https://github.com/OpenRCT2/OpenRCT2/pull/2328

**Reply by SequoiaDu** [reply]
Good catch! The new methods can be run on the GPU and I have not seen performance issues. AMD's FSR is a particularly interesting one to me.

**Comment by toimine** [comment]
You can distribute a program that upscales graphics, then make the user take their rct2 and put it through the program to get better graphics, then you won't have legal issues

**Comment by J-PIE-314** [comment]
This request is similar to #16884.
